src/test/ui/impl-trait/dyn-trait-return-should-be-impl-trait.stderr

```diff
@@ -95,6 +107,18 @@ LL |     }
 LL |     Box::new(42)
    |
 
+error[E0308]: `if` and `else` have incompatible types
```

Comment:
Why is this happening?

Reply:
Because before both arms were being coerced to the common bare trait object, and that is no longer happening after making the big `if let ty::Dynamic(_)`/`check_expr` without expectations hack. I'm ok with that slight worsening of the output to gain the accurate output. If we had a way to get around the need of the hack (like evaluating the type of the tail expression without recording it and carrying it in `traits::SizedReturnType` for example), then we could get rid of the new error.